include/grpc/event_engine/memory_allocator.h

```diff
@@ -0,0 +1,242 @@
+
+  // Release some bytes that were previously reserved.
+  void Release(size_t n) { return allocator_->Release(n); }
+
+  //
+  // The remainder of this type are helper functions implemented in terms of
+  // Reserve/Release.
+  //
+
+  // An automatic releasing reservation of memory.
+  class Reservation {
+   public:
+    Reservation() = default;
+    Reservation(const Reservation&) = delete;
+    Reservation& operator=(const Reservation&) = delete;
+    Reservation(Reservation&&) = default;
+    Reservation& operator=(Reservation&&) = default;
+    ~Reservation() {
+      if (allocator_ != nullptr) allocator_->Release(size_);
+    }
+
+   private:
+    friend class MemoryAllocator;
+    Reservation(std::shared_ptr<BasicMemoryAllocator> allocator, size_t size)
+        : allocator_(allocator), size_(size) {}
+
+    std::shared_ptr<BasicMemoryAllocator> allocator_ = nullptr;
+    size_t size_ = 0;
+  };
+
+  // Reserve bytes from the quota and automatically release them when
+  // Reservation is destroyed.
+  Reservation MakeReservation(MemoryRequest request) {
+    return Reservation(allocator_, Reserve(request));
+  }
+
+  // Allocate a new object of type T, with constructor arguments.
+  // The returned type is wrapped, and upon destruction the reserved memory
+  // will be released to the allocator automatically. As such, T must have a
+  // virtual destructor so we can insert the necessary hook.
+  template <typename T, typename... Args>
+  typename std::enable_if<std::has_virtual_destructor<T>::value, T*>::type New(
+      Args&&... args) {
+    // Wrap T such that when it's destroyed, we can release memory back to the
+    // allocator.
+    class Wrapper final : public T {
+     public:
+      explicit Wrapper(std::shared_ptr<BasicMemoryAllocator> allocator,
+                       Args&&... args)
+          : T(std::forward<Args>(args)...), allocator_(std::move(allocator)) {}
+      ~Wrapper() override { allocator_->Release(sizeof(*this)); }
+
+     private:
+      const std::shared_ptr<BasicMemoryAllocator> allocator_;
+    };
+    Reserve(sizeof(Wrapper));
+    return new Wrapper(allocator_, std::forward<Args>(args)...);
+  }
+
+  // Construct a unique ptr immediately.
+  template <typename T, typename... Args>
+  std::unique_ptr<T> MakeUnique(Args&&... args) {
+    return std::unique_ptr<T>(New<T>(std::forward<Args>(args)...));
+  }
+
+  // Allocate a slice, using MemoryRequest to size the number of returned bytes.
+  // For a variable length request, check the returned slice length to verify
+  // how much memory was allocated.
+  // Takes care of reserving memory for any relevant control structures also.
+  grpc_slice MakeSlice(MemoryRequest request);
+
+  // A C++ allocator for containers of T.
+  template <typename T>
+  class Container {
+   public:
+    using value_type = T;
+
+    // Construct the allocator: \a underlying_allocator is borrowed, and must
```

Comment:
Needs `///` (3 slashes) for `\a` to be meaningful. Should be used throughout. https://www.doxygen.nl/manual/docblocks.html

Reply:
Done.